cirq-core/cirq/devices/noise_utils.py

```diff
@@ -0,0 +1,232 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import TYPE_CHECKING, Any, Dict, Tuple, Type, Union
+import warnings
+import numpy as np
+
+from cirq import ops, protocols, value
+
+if TYPE_CHECKING:
+    import cirq
+
+
+# Tag for gates to which noise must be applied.
+PHYSICAL_GATE_TAG = 'physical_gate'
+
+
+@value.value_equality(distinct_child_types=True)
+class OpIdentifier:
+    """Identifies an operation by gate and (optionally) target qubits."""
+
+    def __init__(self, gate_type: Type['cirq.Gate'], *qubits: 'cirq.Qid'):
+        self._gate_type = gate_type
+        self._gate_family = ops.GateFamily(gate_type)
+        self._qubits: Tuple['cirq.Qid', ...] = tuple(qubits)
+
+    @property
+    def gate_type(self) -> Type['cirq.Gate']:
+        # set to a type during initialization, never modified
+        return self._gate_type
+
+    @property
+    def qubits(self) -> Tuple['cirq.Qid', ...]:
+        return self._qubits
+
+    def _predicate(self, *args, **kwargs):
+        return self._gate_family._predicate(*args, **kwargs)
+
+    def swapped(self):
+        return OpIdentifier(self.gate_type, *self.qubits[::-1])
+
+    def is_proper_subtype_of(self, op_id: 'OpIdentifier'):
+        """Returns true if this is contained within op_id, but not equal to it.
+
+        If this returns true, (x in self) implies (x in op_id), but the reverse
+        implication does not hold. op_id must be more general than self (either
+        by accepting any qubits or having a more general gate type) for this
+        to return true.
+        """
+        more_specific_qubits = self.qubits and not op_id.qubits
+        more_specific_gate = self.gate_type != op_id.gate_type and issubclass(
+            self.gate_type, op_id.gate_type
+        )
+        return (
+            (more_specific_qubits or more_specific_gate)
+            and (more_specific_qubits or self.qubits == op_id.qubits)
+            and (more_specific_gate or self.gate_type == op_id.gate_type)
+        )
+
+    def __contains__(self, item: Union[ops.Gate, ops.Operation]) -> bool:
+        if isinstance(item, ops.Gate):
```

Comment:
None of the models from noise required that, so I opted not to support it. Is there a particular case you have in mind?

(CircuitOperations are a definite complication as far as matching is concerned, but I don't expect them to be used as the match target for an OpIdentifier.)

Reply:
In my use cases I had custom Operations that were not constructed using a Gate.  I suppose the user could always generate a placeholder Gate object to identify any such Operation. 

I can't give any native cirq Operation where this may be an issue since I am not so familiar with all the native cirq ops. One place you may have issues is with measurement operations. I.e. the output of cirq.measure has a gate output that is key dependent.

Reply:
That case should be OK - this only checks against qubits and gate types, both of which are defined explictly in the cirq.measure call. On a more general note, Cirq has been moving towards an "all operations have gates" structure (#4683), which should help with any remaining problem cases.